homeworks/week5/HW5_Khoi.java

```diff
@@ -0,0 +1,85 @@
+import java.util.ArrayList;
+import java.util.List;
+
+Time complexity: O(h) where h is the height of the tree
+Space complexity: O(n) in the worst case, where n is the number of employees
+class Main {
+  public static void main(String[] args) {
+    int[] boss = { 1, 3, 3, -1, -1, 4, 2, 8, 3 };
+    System.out.println(lowestCommonBoss(boss, 1, 1));  // 1
+		System.out.println(lowestCommonBoss(boss, 2, 0));  // 3
+    System.out.println(lowestCommonBoss(boss, 6, 0));  // 3
+    System.out.println(lowestCommonBoss(boss, 7, 1));  // 3
+    System.out.println(lowestCommonBoss(boss, 4, 5));  // 4
+    System.out.println(lowestCommonBoss(boss, 6, 5));  // -1
+    System.out.println(lowestCommonBoss(boss, 4, 2));  // -1
+    System.out.println(lowestCommonBoss(boss, 4, 3));  // -1
+    System.out.println(lowestCommonBoss(boss, 4, -1));  // -1
+    System.out.println(lowestCommonBoss(boss, 3, 7));  // 3
+  }
+
+	static int lowestCommonBoss(int[] boss, int e1, int e2) {
+
+    int result = -100;
+
+    //Edge cases
+    if (e1==e2)  return e1;
+    if ((e1==-1)||(e2==-1)) return -1;
+    
+    //Paths from employees to boss
+    List<Integer> path1 = new ArrayList<>();
+    path1.add(e1);
+    List<Integer> path2 = new ArrayList<>();
```

Comment:
Using extra storage to find path to -1 is a good first approach.

Sample solution provide O(1) space complexity https://repl.it/@HienTo/lowestCommonBoss-Solution#Main.java